libbpf-tools/softirqs.c

```diff
@@ -0,0 +1,277 @@
+	[TASKLET_SOFTIRQ] = "tasklet",
+	[SCHED_SOFTIRQ] = "sched",
+	[HRTIMER_SOFTIRQ] = "hrtimer",
+	[RCU_SOFTIRQ] = "rcu",
+};
+
+static int print_count(int fd)
+{
+	char *units = env.nanoseconds ? "nsecs" : "usecs";
+	__u64 count, zero = 0;
+	__u32 vec;
+	int err;
+
+	printf("%-16s %6s%5s\n", "SOFTIRQ", "TOTAL_", units);
+
+	for (vec = 0; vec < NR_SOFTIRQS; vec++) {
+		err = bpf_map_lookup_elem(fd, &vec, &count);
+		if (err < 0) {
+			fprintf(stderr, "failed to lookup count: %d\n", err);
+			return -1;
+		}
+		if (count > 0)
+			printf("%-16s %11llu\n", vec_names[vec], count);
+	}
+
+	for (vec = 0; vec < NR_SOFTIRQS; vec++) {
+		err = bpf_map_update_elem(fd, &vec, &zero, 0);
+		if (err < 0) {
+			fprintf(stderr, "failed to cleanup count : %d\n", err);
+			return -1;
+		}
+	}
+
+	return 0;
+}
+
+static int print_hist(int fd)
+{
+	char *units = env.nanoseconds ? "nsecs" : "usecs";
+	struct hist hist, zero = {};
+	__u32 vec;
+	int err;
+
+	for (vec = 0; vec < NR_SOFTIRQS; vec++) {
+		err = bpf_map_lookup_elem(fd, &vec, &hist);
+		if (err < 0) {
+			fprintf(stderr, "failed to lookup hist: %d\n", err);
+			return -1;
+		}
+		if (!memcmp(&zero, &hist, sizeof(hist)))
+			continue;
+		printf("softirq = %s\n", vec_names[vec]);
+		print_log2_hist(hist.slots, MAX_SLOTS, units);
+		printf("\n");
+	}
+
+	for (vec = 0; vec < NR_SOFTIRQS; vec++) {
+		err = bpf_map_update_elem(fd, &vec, &zero, 0);
+		if (err < 0) {
+			fprintf(stderr, "failed to cleanup hist : %d\n", err);
+			return -1;
+		}
+	}
+
+	return 0;
+}
+
+int main(int argc, char **argv)
+{
+	static const struct argp argp = {
+		.options = opts,
+		.parser = parse_arg,
+		.doc = argp_program_doc,
+	};
+	struct softirqs_bpf *obj;
+	struct tm *tm;
+	char ts[32];
+	time_t t;
+	int err;
+
+	err = argp_parse(&argp, argc, argv, 0, NULL, NULL);
+	if (err)
+		return err;
+
+	libbpf_set_print(libbpf_print_fn);
+
+	err = bump_memlock_rlimit();
+	if (err) {
+		fprintf(stderr, "failed to increase rlimit: %d\n", err);
+		return 1;
+	}
+
+	obj = softirqs_bpf__open();
+	if (!obj) {
+		fprintf(stderr, "failed to open and/or load BPF object\n");
+		return 1;
+	}
+
+	/* initialize global data (filtering options) */
+	obj->rodata->targ_dist = env.distributed;
+	obj->rodata->targ_ns = env.nanoseconds;
+
+	err = softirqs_bpf__load(obj);
+	if (err) {
+		fprintf(stderr, "failed to load BPF object: %d\n", err);
+		goto cleanup;
+	}
+
+	err = softirqs_bpf__attach(obj);
+	if (err) {
+		fprintf(stderr, "failed to attach BPF programs\n");
+		goto cleanup;
+	}
+
+	signal(SIGINT, sig_handler);
+
+	printf("Tracing soft irq event time... Hit Ctrl-C to end.\n");
+
+	/* main: poll */
+	while (1) {
+		sleep(env.interval);
+		printf("\n");
+
+		if (env.timestamp) {
+			time(&t);
+			tm = localtime(&t);
+			strftime(ts, sizeof(ts), "%H:%M:%S", tm);
+			printf("%-8s\n", ts);
+		}
+
+		if (!env.distributed)
+			err = print_count(bpf_map__fd(obj->maps.counts));
+		else
+			err = print_hist(bpf_map__fd(obj->maps.hists));
```

Comment:
similar to another PR comment, maybe pass `struct bpf_map *` as arguments?

Reply:
Sure,  as we've change bpf arrays to global values, so I change it to 

```c
static int print_hist(struct softirqs_bpf__bss *bss);
static int print_count(struct softirqs_bpf__bss *bss);
```

instead.